addon/utils/violations-helper.ts

```diff
@@ -81,13 +85,13 @@ export class ViolationsHelper {
    * @public
    * @return {Array}
    */
-  filterBy(key, value) {
-    if (key === 'rule') {
+  filterBy(key: keyof Result, value: any) {
+    if (<any>key === 'rule') {
```

Comment:
This (existing code) seems wrong. There is no `rule` key on an `axe.Result`, so I'm not really sure what this code was trying to do.

Reply:
For the curious:

```
This condition will always return 'false' since the types '"description" | "help" | "helpUrl" | "id" | "impact" | "tags" | "nodes"' and '"rule"' have no overlap.ts(2367)
```

Reply:
The `rule` property was replaced by `id` in more recent releases of axe.

Reply:
OK, so if we're on a more recent version, can we just remove this? Why would it be kept if we control the version of `axe-core`?

Reply:
No, this looks like an internal construct, not a direct reference to axe.Result, as this is being referred to in tests.

https://github.com/ember-a11y/ember-a11y-testing/blob/e3ceeb27783b12275b26fa743ac08b685d52f9d6/tests/unit/utils/violations-helper-test.js#L84

https://github.com/ember-a11y/ember-a11y-testing/blob/e3ceeb27783b12275b26fa743ac08b685d52f9d6/tests/unit/utils/violations-helper-test.js#L58

Reply:
Gets aggregated started here: https://github.com/ember-a11y/ember-a11y-testing/blob/e3ceeb27783b12275b26fa743ac08b685d52f9d6/addon-test-support/audit.ts#L42

Reply:
So this is where TS is showing there being rough edges in the API.

https://github.com/ember-a11y/ember-a11y-testing/blob/master/addon-test-support/audit.ts#L50

Shows that a `Result` object is being pushed into this helper. 

Similarly:

https://github.com/ember-a11y/ember-a11y-testing/blob/e254ebad3e23d81384a18bd9b91868770d424dc8/addon/instance-initializers/axe-component.js#L180

Is a of type `axe.Result[]`. I'm not sure why it's being used in a different way elsewhere.

What do you think?

Reply:
OK it looks like this is trying to do two different things in two different places. 

https://github.com/ember-a11y/ember-a11y-testing/blob/e254ebad3e23d81384a18bd9b91868770d424dc8/addon/instance-initializers/axe-component.js#L183

Is pushing `NodeResult` into this helper instead. From what I can tell, either type has a `node` or `rule` property.

Reply:
The answer here may be just to extract these other pieces to a separate addon.

Reply:
I think TS may be a little off. As far as I know, the axe results object has the following basic shape:
```
{
  inapplicable: [],
  passes: [],
  violations: [
    {
       id,
       impact,
       tags: [],
       description,
       help,
       nodes: [
          { }
       ]
    }
  ]
}
```

So what gets pushed into violationsHelper is `results.violations[i]`.

Reply:
And it looks like `node` isn't a member of the violations object, it's just used as a means of looking up the first HTML tag in the nodes array.

https://github.com/ember-a11y/ember-a11y-testing/blob/e3ceeb27783b12275b26fa743ac08b685d52f9d6/addon-test-support/audit.ts#L42

Reply:
Yeah, the way this is done isn't very TypeScript friendly, as `key` can either be a member of the violations object, or it can be `rule`, which is normalized to `id` within this function, and `node`, which gets `violation.nodes[0].target[0]`.

Reply:
Ya this seems like this data structure is being overloaded in a very odd way. Again, I think the way forward is to extract the component-based pieces into a separate addon, and clean up the existing APIs.